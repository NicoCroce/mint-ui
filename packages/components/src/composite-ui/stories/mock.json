{
    "type": "help_center",
    "children": [
        {
            "type": "header"
        },
        {
            "type": "section_title"
        },
        {
            "type": "list",
            "children": [
                {
                    "type": "list_item",
                    "title": "Insurance",
                    "expandRight": true,
                    "url": "/help-center/insurance"
                },
                {
                    "type": "list_item",
                    "title": "Mutual Funds",
                    "expandRight": true,
                    "url": "/help-center/mutual-funds"
                },
                {
                    "type": "list_item",
                    "title": "Gold",
                    "expandRight": true,
                    "url": "/help-center/gold"
                },
                {
                    "type": "list_item",
                    "title": "Loans",
                    "expandRight": true,
                    "url": "/help-center/loans"
                },
                {
                    "type": "list_item",
                    "title": "Something Else",
                    "expandRight": true,
                    "url": "/help-center/something-else"
                }
            ]
        }
    ]
}