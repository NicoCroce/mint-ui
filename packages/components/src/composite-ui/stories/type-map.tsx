import { ListTypeWrapper } from "./type-wrappers/list";
import { Wrapper } from "./type-wrappers/wrapper";
import { renderComponent } from "..";
import { IDict } from "../../__utils/type";
import { ListItem } from "../../list/list-item";

// Search on how to type any part of it fully since it return an exact signature
export const TypeMap: IDict<(Component: any) => any> = {
	list: renderComponent(ListTypeWrapper),
	list_item: renderComponent(ListItem),
	wrapper: renderComponent(Wrapper)
};
