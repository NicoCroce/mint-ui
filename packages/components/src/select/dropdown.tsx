import * as React from "react";
import styled, { css } from "styled-components";
import COLORS from "../__utils/colors";
import { GlobalStyles } from "../app";

export const DropdownWrapper = styled.div`
	${GlobalStyles};
	margin: 0;
	padding: 0;
	list-style: none;
	background-color: ${COLORS.WHITE};
	border-radius: 4px;
	outline: none;
	-webkit-box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
	box-shadow: 0 2px 8px rgba(0, 0, 0, 0.15);
`;

export const List = styled.ul`
	border: 1px solid ${COLORS.BACKGROUND_GREY};
	border-radius: 4px;
	list-style: none;
	margin: 0;
	padding: 0;
	max-height: 250px;
	margin-bottom: 0;
	padding-left: 0;
	overflow: auto;
	outline: none;
`;

export interface OptionProps {
	key?: string | number;
	disabled?: boolean;
	value?: string | number;
	title?: string;
	children?: React.ReactNode;
	className?: string;
	style?: React.CSSProperties;
	onClick?: (event: React.MouseEvent<HTMLLIElement, MouseEvent>) => void;
}

export const StyledOption = styled.li<Pick<OptionProps, "disabled">>`
	display: block;
	padding: 5px 12px;
	overflow: hidden;
	color: ${COLORS.GREY1};
	font-weight: normal;
	line-height: 22px;
	white-space: nowrap;
	text-overflow: ellipsis;
	cursor: pointer;
	-webkit-transition: background 0.3s ease;
	transition: background 0.3s ease;
	${props =>
		props.disabled &&
		css`
			cursor: no-drop;
		`}
	&:first-child {
		color: rgba(0, 0, 0, 0.65);
		font-weight: 600;
		background-color: #fafafa;
	}
	&:hover {
		background-color: #e6f7ff;
	}
`;

export interface SelectedOption {
	value: string | number;
	title: string;
}
export const Option = ({
	key,
	value,
	disabled = false,
	children,
	onClick
}: OptionProps) => {
	return (
		<StyledOption
			key={key}
			value={value}
			disabled={disabled}
			onClick={onClick}
		>
			{children}
		</StyledOption>
	);
};

interface DropdownProps {
	onSelect?: (option: SelectedOption) => void;
	children:
	| React.ComponentElement<any, any>
	| React.ComponentElement<any, any>[];
}

export const Dropdown = ({ onSelect, children }: DropdownProps) => {
	const handleSelect = (event: React.ChangeEvent<HTMLDivElement>) => {
		const title = event.target.innerText || event.target.textContent || "";
		const value = event.target.getAttribute("value") || "";
		if (event.target) {
			onSelect ? onSelect({ value, title }) : null;
		}
	};
	return (
		<DropdownWrapper data-testid="dropdown">
			<List>
				{React.Children.map(
					children,
					(child: React.CElement<any, any>) =>
						React.cloneElement(child, {
							onClick: handleSelect
						})
				)}
			</List>
		</DropdownWrapper>
	);
};

export default Dropdown;
