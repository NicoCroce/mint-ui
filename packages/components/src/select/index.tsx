import * as React from "react";
import styled from "styled-components";
import { Colors } from "@turtlemint/core";
import ExpandMore from "../icons/expand-more";
import ExpandLess from "../icons/expand-less";
import Dropdown, { SelectedOption } from "./dropdown";
import SIZE from "../__utils/size";

export const SelectWrapper = styled.div`
	width: 300px;
	box-sizing: border-box;
`;
export const StyledSelectCTA = styled.div`
	box-sizing: border-box;
	outline: none;
	border: 0;
	padding: 8px 16px;
	border: 1px solid ${Colors.DISABLED};
	border-radius: 4px;
	font-size: ${SIZE.INPUT_FONT_SIZE};
	color: ${Colors.GREY1};
	max-width: 300px;
	&:hover {
		border: 1px solid ${Colors.PERSIAN_GREEN};
		cursor: pointer;
	}
	display: flex;
	justify-content: space-between;
	align-items: center;
`;

const ArrowWrapper = styled.span`
	color: ${Colors.GREY1};
`;

const ArrowToggle = ({ open }: { open: boolean }) => (
	<ArrowWrapper>{!open ? <ExpandMore /> : <ExpandLess />}</ArrowWrapper>
);

interface SelectCTAProps {
	value: string;
	showArrow?: boolean;
	open?: boolean;
	onClick?: () => void;
}

export const SelectCTA = ({
	value,
	showArrow = false,
	open,
	onClick
}: SelectCTAProps) => {
	const handleSelectCTAClick = () => {
		onClick ? onClick() : null;
	};
	return (
		<StyledSelectCTA onClick={handleSelectCTAClick}>
			<span>{value}</span>
			{showArrow ? <ArrowToggle open={open as boolean} /> : null}
		</StyledSelectCTA>
	);
};

interface SelectProps {
	value: string;
	onSelect?: (option: SelectedOption) => void;
	children:
		| React.ComponentElement<any, any>
		| React.ComponentElement<any, any>[];
}

export const Select = ({ value, onSelect, children }: SelectProps) => {
	const [open, setOpen] = React.useState(false);
	const handleSelect = (option: SelectedOption) => {
		setOpen(!open);
		onSelect ? onSelect(option) : null;
	};
	return (
		<SelectWrapper data-testid="select-wrapper">
			<SelectCTA
				open={open}
				value={value}
				showArrow={true}
				onClick={() => setOpen(!open)}
			/>
			{open ? (
				<Dropdown onSelect={handleSelect}>{children}</Dropdown>
			) : null}
		</SelectWrapper>
	);
};

export default Select;
