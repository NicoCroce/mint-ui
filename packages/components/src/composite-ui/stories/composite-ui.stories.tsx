import * as React from "react";
import { storiesOf } from "@storybook/react";
import { CompositeUI } from "../index";
import config from "./mock.json";
import { TypeMap } from "./type-map";

const stories = storiesOf("Composite UI", module);

stories.add("default", () => {
	return <CompositeUI config={config} typeMap={TypeMap} />;
});
