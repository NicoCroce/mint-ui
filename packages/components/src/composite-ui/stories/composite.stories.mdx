import { Meta, Story, Preview, Props } from '@storybook/addon-docs/blocks';
import { CompositeUI } from "../index";
import config from "./mock.json";
import { TypeMap } from "./type-map";

<Meta title="CompositeUI" component={CompositeUI} />

# Composite UI

With `CompositeUI` you can render any nested layout with specified structure in config structure
in a single line


<Preview>
  <Story name="basic">
   <CompositeUI config={config} typeMap={TypeMap} />
  </Story>
</Preview>

### Props Table

<Props of={CompositeUI} />

### Config structure

```
{
    "type": "help_center",
    "children": [
        {
            "type": "header"
        },
        {
            "type": "section_title"
        },
        {
            "type": "list",
            "children": [
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "warning",
                            "color": "blue"
                        },
                        "title": "Insurance"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "close"
                        },
                        "title": "Mutual Funds"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "settings",
                            "color": "grey"
                        },
                        "title": "Gold"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "notification_important",
                            "color": "red"
                        },
                        "title": "Loans"
                    }
                },
                {
                    "type": "list_item",
                    "props": {
                        "avatar": {
                            "icon": "info"
                        },
                        "title": "Something Else"
                    }
                }
            ]
        }
    ]
}
```
