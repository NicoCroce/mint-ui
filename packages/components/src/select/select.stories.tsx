import * as React from "react";
import { storiesOf } from "@storybook/react";
import Select from "./";
import { Option } from "./dropdown";

const stories = storiesOf("Select", module);

export const data = [
    {
        value: 1,
        text: "Item 1"
    },
    {
        value: 2,
        text: "Item 2"
    }
];
interface DemoSelectProps {
    onSelect?: (value: string) => void;
}
export const DemoSelect = ({ onSelect }: DemoSelectProps) => {
    const [value, setValue] = React.useState("Select Value");

    const handleSelect = (option: any) => {
        setValue(option.title);
        //for test case
        onSelect(option.title);
    }
    return (
        <Select value={value} onSelect={handleSelect}>
            {
                data.map((d: any) => (
                    <Option
                        key={d.value}
                        value={d.value}>
                        {d.text}
                    </Option>
                ))
            }
        </Select>
    );
}
stories.add("full functional demo", () => (
    <DemoSelect />
))

stories.add("default", () => (
    <Select value="Test value">
        {
            data.map((d: any) => (
                <Option
                    key={d.value}
                    value={d.value}>
                    {d.text}
                </Option>
            ))
        }
    </Select>
))