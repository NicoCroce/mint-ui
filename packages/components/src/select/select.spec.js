import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import Dropdown, { Option } from "./dropdown";
import Select from "./index";
import { DemoSelect } from "./select.stories";

export const data = [
	{
		value: 1,
		text: "Item 1"
	},
	{
		value: 2,
		text: "Item 2"
	}
];
afterEach(cleanup);

describe("Dropdown test suite", () => {
	test("Dropdown snapshot", () => {
		const { asFragment } = render(
			<Dropdown>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Dropdown>
		);
		expect(asFragment()).toMatchSnapshot();
	});
	test("Dropdown renders correctly", () => {
		const { getByTestId } = render(
			<Dropdown>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Dropdown>
		);
		expect(
			getByTestId("dropdown").getElementsByTagName("li")[0].textContent
		).toEqual("Item 1");
	});
	test("Item click works fine", () => {
		let selectedValue = null;
		const handleSelect = option => {
			selectedValue = option;
		};
		const { getByTestId } = render(
			<Dropdown onSelect={handleSelect}>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Dropdown>
		);
		fireEvent.click(getByTestId("dropdown").getElementsByTagName("li")[0]);
		expect(selectedValue.text).toBe("Item 1");
	});
});

describe("Select test suites", () => {
	test("Select label renders correctly", () => {
		const { getByTestId } = render(
			<Select onSelect={() => {}}>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Select>
		);
		expect(
			getByTestId("select-wrapper").getElementsByTagName("div")[0]
		).toBeInTheDocument();
	});
	test("Select snapshot", () => {
		const { asFragment } = render(
			<Select onSelect={() => {}}>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Select>
		);
		expect(asFragment()).toMatchSnapshot();
	});
	test("Select label opens the dropdown", () => {
		const { getByTestId } = render(
			<Select onSelect={() => {}}>
				{data.map(d => (
					<Option key={d.value} value={d.value}>
						{d.text}
					</Option>
				))}
			</Select>
		);
		fireEvent.click(
			getByTestId("select-wrapper").getElementsByTagName("div")[0]
		);
		expect(getByTestId("dropdown")).toBeInTheDocument();
	});

	test("Select dropdown item click sets the selected option's title to label", () => {
		let selectedItem = null;
		const handleSelect = item => {
			selectedItem = item;
		};
		const { getByTestId } = render(<DemoSelect onSelect={handleSelect} />);
		const selectLabel = getByTestId("select-wrapper").getElementsByTagName(
			"div"
		)[0];
		fireEvent.click(selectLabel);

		const liElement = getByTestId("select-wrapper")
			.getElementsByTagName("div")[1]
			.getElementsByTagName("li")[1];
		fireEvent.click(liElement);
		// console.log(selectLabel.getElementsByTagName("span")[0].textContent);
		// expect(getByTestId("select-wrapper").getElementsByTagName("div")[1]).not.toBeInTheDocument();
		expect(selectLabel.getElementsByTagName("span")[0].textContent).toEqual(
			selectedItem.text
		);
	});
});
