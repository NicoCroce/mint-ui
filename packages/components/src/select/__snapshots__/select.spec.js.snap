// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown test suite Dropdown snapshot 1`] = `
<DocumentFragment>
  <div
    class="dropdown__DropdownWrapper-sc-4mjgs1-0 jNYjVA"
    data-testid="dropdown"
  >
    <ul
      class="dropdown__List-sc-4mjgs1-1 dWFdKM"
    >
      <li
        class="dropdown__StyledOption-sc-4mjgs1-2 iFCPBi"
        value="item1"
      >
        Item 1
      </li>
      <li
        class="dropdown__StyledOption-sc-4mjgs1-2 iFCPBi"
        value="item2"
      >
        Item 2
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Select test suites Select snapshot 1`] = `
<DocumentFragment>
  <div
    class="select__SelectWrapper-sc-1trodvu-0 jtVoAg"
    tabindex="0"
  >
    <div
      class="select__StyledSelectCTA-sc-1trodvu-1 bFMGxn"
    >
      <span
        style="color: rgb(232, 232, 232);"
      />
      <svg
        fill="#6c7375"
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <path
          d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
        />
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
